monai/data/box_utils.py

```diff
@@ -0,0 +1,565 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+from abc import ABC, abstractmethod
+from copy import deepcopy
+from typing import Dict, Sequence, Tuple, Type, Union
+
+import numpy as np
+import torch
+
+from monai.config.type_definitions import NdarrayOrTensor
+from monai.utils import look_up_option
+from monai.utils.enums import BoxModeName
+from monai.utils.type_conversion import convert_data_type, convert_to_dst_type
+
+# We support 2-D or 3-D bounding boxes
+SUPPORTED_SPATIAL_DIMS = [2, 3]
+
+
+# TO_REMOVE = 0.0 if the bottom-right corner pixel/voxel is not included in the box,
+#      i.e., when xmin=1., xmax=2., we have w = 1.
+# TO_REMOVE = 1.0  if the bottom-right corner pixel/voxel is included in the box,
+#       i.e., when xmin=1., xmax=2., we have w = 2.
+# Currently, only `TO_REMOVE = 0.0` is supported
+TO_REMOVE = 0.0  # xmax-xmin = w -TO_REMOVE.
+
+
+class BoxMode(ABC):
+    """
+    An abstract class of a ``BoxMode``.
+    A BoxMode is callable that converts box mode of boxes.
+    It always creates a copy and will not modify boxes in place.
+
```

Comment:
I think it'll be great to put some summary of MONAI's bounding box mode in the docstring (either at the beginning of this module or in this class).

For example, we should highlight that this utility module mainly supports rectangular bounding boxes with a few different parameterizations and methods for converting between them. It provides reliable access to the spatial coordinates of the box vertices in the "canonical ordering" -- (x_start, y_start, x_end, y_end) for 2D and (x_start, y_start, z_start, x_end, y_end, z_end) for 3D.  We currently define this ordering as `monai.data.box_utils.StandardMode` and the rest of the detection pipelines mainly assumes boxes in `StandardMode`.